src/use-collection.ts

```diff
@@ -9,19 +9,60 @@ import { useCollectionState } from './use-collection-state.js';
 export function useCollection<T>(allItems: ReadonlyArray<T>, options: UseCollectionOptions<T>): UseCollectionResult<T> {
   const collectionRef = useRef<CollectionRef>(null);
   const [state, actions] = useCollectionState(options, collectionRef);
-  const { items, allPageItems, pagesCount, filteredItemsCount, actualPageIndex } = processItems(
+  const { items, allPageItems, pagesCount, filteredItemsCount, actualPageIndex, itemsTree } = processItems(
     allItems,
     state,
     options
   );
+
+  const expandedItemsSet = new Set<string>();
+  if (options.expandableRows) {
+    for (const item of state.expandedItems) {
+      expandedItemsSet.add(options.expandableRows.getId(item));
+    }
+  }
+
+  let visibleItems = items;
+  if (options.expandableRows) {
+    const flatItems = new Array<T>();
+    const getId = options.expandableRows.getId;
+    const traverse = (items: readonly T[]) => {
+      for (const item of items) {
+        flatItems.push(item);
+        if (expandedItemsSet.has(getId(item))) {
+          traverse(itemsTree.getChildren(item));
+        }
+      }
+    };
+    traverse(items);
+    visibleItems = flatItems;
+  }
+
+  // Removing selected items that are no longer present in items unless keepSelection=true.
   if (options.selection && !options.selection.keepSelection) {
-    const newSelectedItems = processSelectedItems(items, state.selectedItems, options.selection.trackBy);
+    const newSelectedItems = processSelectedItems(visibleItems, state.selectedItems, options.selection.trackBy);
     if (!itemsAreEqual(newSelectedItems, state.selectedItems, options.selection.trackBy)) {
       // This is a recommended pattern for how to handle the state, dependent on the incoming props
       // https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
       actions.setSelectedItems(newSelectedItems);
     }
   }
+
+  // Removing expanded items that are no longer present in items.
+  useEffect(() => {
+    if (options.expandableRows) {
+      const newExpandedItems = new Array<T>();
```

Comment:
Can't we get rid of the for loop with just

```
const newExpandedItems = visibleItems.filter(item => expandedItemsSet.has(options.expandableRows.getId(item)))
```

Reply:
Thanks, makes sense. I will also change how the items are compared to match very similar code for selection state above.